File: objc/objcfind/Sources/objcfind/ArgTokenizer.m
#import "common.h"
#import "ArgTokenizer.h"
#import "FileUtil.h"

@interface ArgTokenizer ()
// private properties
@property NSDictionary<NSString*,NSString*> *boolDict;
@property NSDictionary<NSString*,NSString*> *stringDict;
@property NSDictionary<NSString*,NSString*> *intDict;

@end

@implementation ArgTokenizer

- (instancetype) initWithBoolDict:(NSDictionary<NSString*,NSString*>*)boolDict stringDict:(NSDictionary<NSString*,NSString*>*)stringDict intDict:(NSDictionary<NSString*,NSString*>*)intDict {
    self = [super init];
    if (self) {
        self.boolDict = boolDict;
        self.stringDict = stringDict;
        self.intDict = intDict;
    }
    return self;
}

- (NSArray<ArgToken*> *) tokenizeArgs:(NSArray<NSString*> *)args error:(NSError **)error {
    NSMutableArray *argTokens = [[NSMutableArray alloc] init];

    int i = 1;
    while (i < [args count]) {
        if (*error) {
            return nil;
        }
        NSString *arg = args[i];
        if ([arg hasPrefix:@"-"]) {
            NSMutableArray *argNames = [NSMutableArray array];
            NSString *argVal = nil;
            if ([arg hasPrefix:@"--"]) {
                if ([arg length] > 2) {
                    // Process long arg
                    arg = [arg substringFromIndex:2];
                    NSArray *parts = [arg componentsSeparatedByString:@"="];
                    if ([parts count] > 0) {
                        arg = [parts objectAtIndex:0];
                    }
                    if ([parts count] > 1) {
                        argVal = [parts objectAtIndex:1];
                    }
                    [argNames addObject:arg];
                } else {
                    setError(error, [NSString stringWithFormat:@"Invalid option: %@", arg]);
                    return nil;
                }
            } else if ([arg length] > 1) {
                // Process short arg(s)
                arg = [arg substringFromIndex:1];
                for (NSUInteger i = 0; i < [arg length]; i++) {
                    unichar c = [arg characterAtIndex:i];
                    NSString *cs = [NSString stringWithFormat:@"%C", c];
                    if (self.boolDict[cs]) {
                        [argNames addObject:self.boolDict[cs]];
                    } else if (self.stringDict[cs]) {
                        [argNames addObject:self.stringDict[cs]];
                    } else if (self.intDict[cs]) {
                        [argNames addObject:self.intDict[cs]];
                    } else {
                        setError(error, [NSString stringWithFormat:@"Invalid option: %@", cs]);
                        return nil;
                    }
                }
            } else {
                setError(error, [NSString stringWithFormat:@"Invalid option: %@", arg]);
                return nil;
            }

            for (NSString *argName in argNames) {
                if (self.boolDict[argName]) {
                    [argTokens addObject:[[ArgToken alloc] initWithName:argName type:ArgTokenTypeBool value:@(YES)]];
                } else {
                    if (argVal == nil) {
                        if ([args count] > i+1) {
                            argVal = args[i+1];
                            i++;
                        } else {
                            setError(error, [NSString stringWithFormat:@"Missing argument for option %@", arg]);
                            return nil;
                        }
                    }
                    if (self.stringDict[argName] || [argName isEqualToString:@"settings-file"]) {
                        [argTokens addObject:[[ArgToken alloc] initWithName:argName type:ArgTokenTypeStr value:argVal]];
                    } else if (self.intDict[argName]) {
                        [argTokens addObject:[[ArgToken alloc] initWithName:argName type:ArgTokenTypeInt value:@([argVal integerValue])]];
                    } else {
                        setError(error, [NSString stringWithFormat:@"Invalid option: %@", arg]);
                        return nil;
                    }
                }
            }

        } else {
            [argTokens addObject:[[ArgToken alloc] initWithName:@"path" type:ArgTokenTypeStr value:arg]];
        }
        i++;
    }
    
    return [NSArray arrayWithArray:argTokens];
}

- (NSArray<ArgToken*> *) tokenizeDictionary:(NSDictionary *)dictionary error:(NSError **)error {
    NSMutableArray *argTokens = [[NSMutableArray alloc] init];
    // keys are sorted so that output is consistent across all versions
    NSArray<NSString*> *keys = [[dictionary allKeys] sortedArrayUsingSelector:@selector(compare:)];
    for (NSString *key in keys) {
        NSObject *val = dictionary[key];
        if (self.boolDict[key]) {
            if ([val isKindOfClass:[NSNumber class]]) {
                NSNumber *num = (NSNumber *)val;
                BOOL b = [num boolValue];
                [argTokens addObject:[[ArgToken alloc] initWithName:key type:ArgTokenTypeBool value:@(b)]];
            } else {
                setError(error, [@"Invalid value for option: " stringByAppendingString:key]);
                return nil;
            }
        } else if (self.stringDict[key] || [key isEqualToString:@"settings-file"]) {
            if ([val isKindOfClass:[NSString class]]) {
                NSString *s = (NSString *)val;
                [argTokens addObject:[[ArgToken alloc] initWithName:key type:ArgTokenTypeStr value:s]];
            } else if ([val isKindOfClass:[NSArray class]]) {
                NSArray *arr = (NSArray *)val;
                for (NSObject *o in arr) {
                    if ([o isKindOfClass:[NSString class]]) {
                        NSString *s = (NSString *)o;
                        [argTokens addObject:[[ArgToken alloc] initWithName:key type:ArgTokenTypeStr value:s]];
                    } else {
                        setError(error, [@"Invalid value for option: " stringByAppendingString:key]);
                        return nil;
                    }
                }
            } else {
                setError(error, [@"Invalid value for option: " stringByAppendingString:key]);
                return nil;
            }
        } else if (self.intDict[key]) {
            if ([val isKindOfClass:[NSNumber class]]) {
                NSNumber *num = (NSNumber *)val;
                NSInteger i = [num integerValue];
                [argTokens addObject:[[ArgToken alloc] initWithName:key type:ArgTokenTypeInt value:@(i)]];
            } else {
                setError(error, [@"Invalid value for option: " stringByAppendingString:key]);
                return nil;
            }
        } else {
            setError(error, [NSString stringWithFormat:@"Invalid option: %@", key]);
            return nil;
        }
    }
    return [NSArray arrayWithArray:argTokens];
}

- (NSArray<ArgToken*> *) tokenizeData:(NSData *)data error:(NSError **)error {
    if (NSClassFromString(@"NSJSONSerialization")) {
        id jsonObject = [NSJSONSerialization
                         JSONObjectWithData:data
                         options:0
                         error:error];
        if (*error) {
            setError(error, @"Unable to parse json");
        }
        
        if (![jsonObject isKindOfClass:[NSDictionary class]]) {
            setError(error, @"Invalid json");
        }
        
        return [self tokenizeDictionary:jsonObject error:error];
    } else {
        setError(error, @"Unable to parse json");
    }
    return nil;
}

- (NSArray<ArgToken*> *) tokenizeFile:(NSString *)filePath error:(NSError **)error {
    NSString *expandedPath = [FileUtil expandPath:filePath];
    if (![[NSFileManager defaultManager] fileExistsAtPath:expandedPath]) {
        setError(error, [@"Settings file not found: " stringByAppendingString:filePath]);
        return nil;
    }
    if (![expandedPath hasSuffix:@".json"]) {
        setError(error, [@"Invalid settings file (must be JSON): " stringByAppendingString:filePath]);
        return nil;
    }
    NSData *data = [NSData dataWithContentsOfFile:expandedPath];
    return [self tokenizeData:data error:error];
}

@end

File: swift/swiftfind/Sources/swiftfind/ArgTokenizer.swift
//
//  ArgParser.swift
//  swiftfind
//
//  Created by Cary Clark on 9/1/25.
//  Copyright (c) 2025 Cary Clark. All rights reserved.
//

import Foundation

public enum ArgTokenType {
    case bool, str, int, long
}

public struct ArgToken {
    let name: String
    let type: ArgTokenType
    let value: Any
}

public class ArgTokenizer {
    private var boolDict: [String: String] = [:]
    private var stringDict: [String: String] = [:]
    private var intDict: [String: String] = [:]
    private var longDict: [String: String] = [:]

    public init(_ boolDict: [String: String], _ stringDict: [String: String], _ intDict: [String: String], _ longDict: [String: String]) {
        self.boolDict = boolDict
        self.stringDict = stringDict
        self.intDict = intDict
        self.longDict = longDict
    }

    public func tokenizeArgs(_ args: [String]) throws -> [ArgToken] {
        var i = 0
        var argTokens: [ArgToken] = []
        // default printFiles to true since running as cli
        while i < args.count {
            var arg = args[i]
            if arg.hasPrefix("-") {
                var argNames: [String] = []
                var argVal: String? = nil
                if arg.hasPrefix("--") {
                    if arg.count > 2 {
                        // Process long arg
                        arg = String(arg[arg.index(arg.startIndex, offsetBy: 2)...])
                        if arg.contains("=") {
                            let parts = arg.split(separator: "=")
                            if parts.count > 0 {
                                arg = String(parts[0])
                            }
                            if parts.count > 1 {
                                argVal = String(parts[1])
                            }
                        }
                        argNames.append(arg)
                    } else {
                        throw FindError(msg: "Invalid option: \(arg)")
                    }
                } else if arg.count > 1 {
                    // Process short arg(s)
                    arg = String(arg[arg.index(arg.startIndex, offsetBy: 1)...])
                    for c in arg {
                        let cs = String(c)
                        if boolDict.index(forKey: cs) != nil {
                            argNames.append(boolDict[cs]!)
                        } else if stringDict.index(forKey: cs) != nil {
                            argNames.append(stringDict[cs]!)
                        } else if intDict.index(forKey: cs) != nil {
                            argNames.append(intDict[cs]!)
                        } else if longDict.index(forKey: cs) != nil {
                            argNames.append(longDict[cs]!)
                        } else {
                            throw FindError(msg: "Invalid option: \(cs)")
                        }
                    }
                } else {
                    throw FindError(msg: "Invalid option: \(arg)")
                }
                
                for argName in argNames {
                    if boolDict.index(forKey: argName) != nil {
                        argTokens.append(ArgToken(name: argName, type: ArgTokenType.bool, value: true))
                    } else {
                        if argVal == nil {
                            if args.count > i + 1 {
                                i += 1
                                argVal = args[i]
                            } else {
                                throw FindError(msg: "Missing argument for option \(arg)")
                            }
                        }
                        if stringDict.index(forKey: argName) != nil {
                            argTokens.append(ArgToken(name: argName, type: ArgTokenType.str, value: argVal!))
                        } else if intDict.index(forKey: argName) != nil {
                            let intVal = Int32(argVal!) ?? 0
                            argTokens.append(ArgToken(name: argName, type: ArgTokenType.int, value: intVal))
                        } else if longDict.index(forKey: argName) != nil {
                            let longVal = UInt64(argVal!) ?? 0
                            argTokens.append(ArgToken(name: argName, type: ArgTokenType.long, value: longVal))
                        } else if argName == "settings-file" {
                            argTokens.append(ArgToken(name: argName, type: ArgTokenType.str, value: argVal!))
                        } else {
                            throw FindError(msg: "Invalid option: \(arg)")
                        }
                    }
                }
            } else {
                argTokens.append(ArgToken(name: "path", type: ArgTokenType.str, value: arg))
            }
            i += 1
        }
        return argTokens
    }

    public func tokenizeArgMap(_ argMap: [String: Any]) throws -> [ArgToken] {
        var argTokens: [ArgToken] = []
        // keys are sorted so that output is consistent across all versions
        let keys = argMap.keys.sorted()
        for key in keys {
            if boolDict.index(forKey: key) != nil {
                let value = argMap[key]
                if let bool = value as? Bool {
                    argTokens.append(ArgToken(name: key, type: ArgTokenType.bool, value: bool))
                } else {
                    throw FindError(msg: "Invalid value for option: \(key)")
                }
            } else if stringDict.index(forKey: key) != nil || key == "settings-file" {
                let value = argMap[key]
                if let string = value as? String {
                    argTokens.append(ArgToken(name: key, type: ArgTokenType.str, value: string))
                } else if let stringArray = value as? [String] {
                    for s in stringArray {
                        argTokens.append(ArgToken(name: key, type: ArgTokenType.str, value: s))
                    }
                } else {
                    throw FindError(msg: "Invalid value for option: \(key)")
                }
            } else if intDict.index(forKey: key) != nil {
                let value = argMap[key]
                if let intVal = value as? Int32 {
                    argTokens.append(ArgToken(name: key, type: ArgTokenType.int, value: intVal))
                } else {
                    throw FindError(msg: "Invalid value for option: \(key)")
                }
            } else if longDict.index(forKey: key) != nil {
                let value = argMap[key]
                if let longVal = value as? UInt64 {
                    argTokens.append(ArgToken(name: key, type: ArgTokenType.long, value: longVal))
                } else {
                    throw FindError(msg: "Invalid value for option: \(key)")
                }
            } else {
                throw FindError(msg: "Invalid option: \(key)")
            }
        }
        return argTokens
    }

    public func tokenizeJson(_ jsonString: String) throws -> [ArgToken] {
        var argTokens: [ArgToken] = []
        do {
            if let json = try JSONSerialization.jsonObject(with: jsonString.data(using: .utf8)!,
                                                           options: []) as? [String: Any]
            {
                argTokens = try tokenizeArgMap(json)
            }
        } catch let error as FindError {
            throw error
        } catch let error {
            throw FindError(msg: "Failed to load: \(error.localizedDescription)")
        }
        return argTokens
    }

    public func tokenizeFile(_ filePath: String) throws -> [ArgToken] {
        var argTokens: [ArgToken] = []
        let expandedPath = FileUtil.expandPath(filePath)
        if !FileUtil.exists(expandedPath) {
            throw FindError(msg: "Settings file not found: \(filePath)")
        }
        if !expandedPath.hasSuffix(".json") {
            throw FindError(msg: "Invalid settings file (must be JSON): \(filePath)")
        }
        do {
            let fileUrl = URL(fileURLWithPath: expandedPath)
            let jsonString = try String(contentsOf: fileUrl, encoding: .utf8)
            argTokens = try tokenizeJson(jsonString)
        } catch let error as FindError {
            throw error
        } catch let error {
            throw FindError(msg: "Failed to load: \(error.localizedDescription)")
        }
        return argTokens
    }
}
